### ISSUE TITLE
Is there a way to access multiple claims when using custom authentication?

### ISSUE THREAD

**Issue by TellonUK** [Azure/static-web-apps#429]
I am using Azure B2C as my identity provider and it returns the name by default (firstname givename). I can access the user's email address if I use the below configuration, but I then no longer have the users name.

`"nameClaimType": "emails",`

Is there a way to request multiple claims from the decoded token? In my situation name and emails ?

**Comment by mkarmark** [comment]
There currently is not a way to do that. In what form would you want multiple claims presented to you? Would a comma separated list make sense here?

**Comment by TellonUK** [comment]
I think that makes sense. Adding something like the below to the config, and then receiving a response in the same order.

Config

optionalClaims : [ 'name', 'emails', 'etc']

Additional to client provider response

optionalClaims : [ 'firstname givenname', 'xyz@abc.com', '...']

If a claims is requested that isn't in the jwt response, then the same as you have setup scopes would be ideal, a simple 500 error.

**Comment by GaryParr** [comment]
OK, so... perhaps I'm missing something but why can't we have access to the whole blasted B2C claim?  I understand SWA needs a single value to map into their existing clientPrincipal.userDetails property... but what's wrong with adding a clientPrincipal.providerClaim field that contains everything?  This would assist with other efforts as well, such as https://github.com/Azure/static-web-apps/issues/415

**Comment by mrfoster** [comment]
I think this is absolutely necessary and was surprised to find that I wasn't able to get at them. I'm sure there are many many reasons to do so, but I am storing the plan the user has subscribed to in a claim and want to be able to retrieve it..

**Comment by alefranz** [comment]
Hello! I'm currently looking to migrate a website from AppService to Static Web Apps and I am blocked by this.
I think the ability to override the json object passed as `claims` parameter would be enough to start with, as well as then adding these to the client principal data.

**Comment by anthonychu** [comment]
@alefranz We're working on some improvements in this area. Can you describe a bit more about what you want to do?

**Comment by alefranz** [comment]
> @alefranz We're working on some improvements in this area. Can you describe a bit more about what you want to do?

Hi @anthonychu , the OIDC provider I am using allow me to provide multiple claims so to be able to request for more user info from the UserInfo endpoint e.g. profile picture. Currently I am using the ASP.NET Core OpenIdConnect library and I override the `OnRedirectToIdentityProvider` to replace the claims parameter with the JSON with the list of claims I need. Is not the most elegant solution but allows me to achieve this, so it would be nice to have a similar solution here as well. Let me know if you need more details.

**Comment by anthonychu** [comment]
@miwebst Can you see if it’s possible to support this?

**Comment by danwatford** [comment]
I raised a related issue against the documentation as a I didn't realise this repository existed for reporting problems - https://github.com/MicrosoftDocs/azure-docs/issues/86803

Per the linked issue, claims are now available at the /.auth/me endpoint, which is good, but documentation suggests they should also be available to functions via the x-ms-client-principal header.

@anthonychu , do you know if claims likely to be included in the x-ms-client-principal header in the future?

**Comment by gordonk** [comment]
> Per the linked issue, claims are now available at the /.auth/me endpoint, which is good, but documentation suggests they should also be available to functions via the x-ms-client-principal header.
> 
> @anthonychu , do you know if claims likely to be included in the x-ms-client-principal header in the future?

Yes thanks for all your hard work, but like Dan above I have not seen claims passed in the x-ms-client-principal header even thought the documentation clearly states they are.
